src/persist-client/src/impl/trace.rs

```diff
@@ -0,0 +1,1074 @@
+
+use std::fmt::Debug;
+
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::trace::Description;
+use timely::progress::frontier::AntichainRef;
+use timely::progress::{Antichain, Timestamp};
+
+use crate::r#impl::state::HollowBatch;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct FueledMergeReq<T> {
+    pub desc: Description<T>,
+    pub inputs: Vec<HollowBatch<T>>,
+}
+
+#[derive(Debug)]
+pub struct FueledMergeRes<T> {
+    pub output: HollowBatch<T>,
+}
+
+/// An append-only collection of compactable update batches.
+///
+/// In an effort to keep our fork of Spine as close as possible to the original,
+/// we push as many changes as possible into this wrapper.
+#[derive(Debug, Clone)]
+pub struct Trace<T> {
+    spine: Spine<T>,
+    merge_reqs: Vec<FueledMergeReq<T>>,
+}
+
+impl<T: Timestamp + Lattice> Default for Trace<T> {
+    fn default() -> Self {
+        Self {
+            spine: Spine::new(),
+            merge_reqs: Vec::new(),
+        }
+    }
+}
+
+impl<T: Timestamp + Lattice> Trace<T> {
+    pub fn since(&self) -> &Antichain<T> {
+        &self.spine.since
+    }
+
+    pub fn upper(&self) -> &Antichain<T> {
+        &self.spine.upper
+    }
+
+    pub fn downgrade_since(&mut self, since: Antichain<T>) {
+        self.spine.since = since;
+    }
+
+    pub fn push_batch(&mut self, batch: HollowBatch<T>) {
+        self.spine
+            .insert(SpineBatch::Merged(batch), &mut self.merge_reqs);
+    }
+
+    pub fn map_batches<F: FnMut(&HollowBatch<T>)>(&self, mut f: F) {
+        self.spine.map_batches(move |b| match b {
+            SpineBatch::Merged(b) => f(b),
+            SpineBatch::Fueled { parts, .. } => {
+                for b in parts.iter() {
+                    f(b);
+                }
+            }
+        })
+    }
+
+    pub fn take_merge_reqs(&mut self) -> Vec<FueledMergeReq<T>> {
+        std::mem::take(&mut self.merge_reqs)
+    }
+
+    pub fn apply_merge_res(&mut self, res: &FueledMergeRes<T>) -> bool {
+        let maybe_replace = |b: &mut SpineBatch<T>| -> bool {
+            let (desc, _parts) = match b {
+                SpineBatch::Merged(_) => return false,
+                SpineBatch::Fueled { desc, parts } => (desc, parts),
+            };
+            // We could be more lenient here and swap out a subset if the descs
+            // line up. See if this matters in practice before we add the
+            // complexity.
+            if desc != &res.output.desc {
+                return false;
+            }
+
+            *b = SpineBatch::Merged(res.output.clone());
+            return true;
+        };
+
+        let mut applied = false;
+        for batch in self.spine.merging.iter_mut().rev() {
+            if applied {
+                break;
+            }
+            match batch {
+                MergeState::Double(MergeVariant::InProgress(batch1, batch2, m)) => {
+                    // WIP make the two batch clones share internals
```

Comment:
thoughts? this is something we'd need to do anyway to reuse the real Spine. I'm thinking something like this (with better names)

```
struct TraceBatch<T> {
   desc: Description<T>
   len: usize,
   core: Arc<Mutex<TraceBatchCore<T>>>,
}
enum TraceBatchCore<T> {
    Merged(HollowBatch<T>)
    Fueled(Vec<HollowBatch<T>>)
}
```

to make things work with real spine, we could register an Arc of the "core" of each batch as it's inserted into the spine. then apply_merge_res could mutate it via the one in the registry. this structure has the advantage that the desc and len never change out from under Spine. this is not true for len as things are typed right now, but it seems like a good property to have. after swapping a merge res in, the len of the TraceBatch and the one of the core might disagree, but this is fine since the former is an upper bound on the latter. we'd have to figure out how to remove things from the registry once spine is finished with them, maybe something with the spine logger or using a weak Arc in the registry?

Reply:
indeed we do have to update the ones in the FuelingMerge because when that merge finishes, those are the ones that get turned into the new batch (this was a bug I had to track down at some point before this was even a PR 😅 )

https://github.com/MaterializeInc/materialize/pull/13160/files#diff-d74ca1805afb882373ab284ccd11e8eedda43152b4f31d7f309941a3b894d3c9R293-R334

the code we had before was fine without the Arc+SpineBatchCore, we only need this new thing if we want to work with regular Spine. i'm pretty divided on which of the two we should start with. thoughts?

Reply:
The diff in Spine that we need when we don't have `Arc+SpineBatchCore` is that we need to patch things up "outside" the batches, right? I think I like that slightly more, because it doesn't rely on the somewhat hidden shared nature of `SpineBatch` for correctness. But I'd also like it if we stayed true to vanilla `Spine` as much as possible. Though call!